api/services/locations.js

```diff
@@ -0,0 +1,170 @@
+import config from 'config';
+import AWS from 'aws-sdk';
+import { has, sortBy } from 'lodash';
+import { db } from './db';
+
+// Athena config
+const {
+  accessKeyId,
+  secretAccessKey,
+  OutputLocation,
+  Database,
+  region,
+  measurementsTable
+} = config.get('athena');
+
+// Athena client
+const athena = new AWS.Athena({
+  accessKeyId,
+  secretAccessKey,
+  region
+});
+
+// The locations query
+const QueryString = `
+  SELECT
+    UPPER(country) as country,
+    round(coordinates.longitude, 5) AS lon,
+    round(coordinates.latitude, 5) AS lat,
+    min(date.local) AS firstUpdated,
+    max(date.local) AS lastUpdated
+  FROM ${measurementsTable}
+  GROUP BY
+    UPPER(country),
+    round(coordinates.longitude,5),
+    round(coordinates.latitude,5)
+  ORDER BY
+    UPPER(country), 
+    firstUpdated
+`;
+
+// Flow control variables
+const checkQueryStatusInterval = 2000;
+let intervalId = false;
+let QueryExecutionId;
+
+const checkQueryStatus = async function () {
+  function stopListener () {
+    intervalId = clearInterval(intervalId);
+  }
+
+  try {
+    const data = await athena.getQueryExecution({ QueryExecutionId }).promise();
+
+    // If response is malformed, stop listener.
+    if (!has(data, 'QueryExecution.Status.State')) {
+      console.log('Error while fetching Athena query status.');
+      stopListener();
+      return;
+    }
+
+    // Process state
+    const { State } = data.QueryExecution.Status;
+    if (State === 'SUCCEEDED') {
+      // update locations
+      console.log('Athena query was finished successfully!');
+      stopListener();
+      upsertLocations();
+    } else if (State !== 'RUNNING') {
+      console.log('Athena query was interrupted.');
+      stopListener();
+    } else {
+      console.log('Athena query is running...');
+    }
+  } catch (err) {
+    // Unexpected error occurred.
+    console.log(err, err.stack);
+    stopListener();
+  }
+};
+
+export const fetchLocations = async function () {
+  // Check if a query is already running
+  if (!intervalId) {
+    console.log('Creating Athena query.');
+    const data = await athena
+      .startQueryExecution({
+        QueryString,
+        QueryExecutionContext: {
+          Database
+        },
+        ResultConfiguration: {
+          OutputLocation
+        }
+      })
+      .promise();
+
+    // Register query run and start listener
+    QueryExecutionId = data.QueryExecutionId;
+    intervalId = setInterval(checkQueryStatus, checkQueryStatusInterval);
+  }
+};
+
+export const upsertLocations = async function () {
+  let queryResults = await athena
+    .getQueryResults({ QueryExecutionId })
+    .promise();
+
+  if (has(queryResults, 'ResultSet.Rows')) {
+    let rows = queryResults.ResultSet.Rows;
```

Comment:
Sorry if I am missing the logic to do this, but do you know if this will get **all** the results or does it need to be looped or paged to get all the results?